File: src/forms/form_selects.py
"""
Extraction utilities for select dropdown controls.
"""
from forms.utils.form_extract_common import find_label_for
from forms.utils.logger import get_logger

logger = get_logger('form_extract_selects')

def extract_selects(container):
    """Extract select dropdowns and their options"""
    selects = container.find_all('select')
    
    result = []
    for select in selects:
        options = []
        for option in select.find_all('option'):
            options.append({
                'value': option.get('value', ''),
                'text': option.text.strip(),
                'selected': option.has_attr('selected')
            })
        
        select_info = {
            'id': select.get('id', ''),
            'name': select.get('name', ''),
            'required': select.has_attr('required'),
            'disabled': select.has_attr('disabled'),
            'options': options
        }
        
        # Try to find label
        select_info['label'] = find_label_for(container, select)
        
        # Check for Chosen enhancement - escape selector to handle special characters
        select_id = select.get('id', '')
        if select_id:
            try:
                # Use a safer approach to find the chosen container
                chosen_class = f"{select_id}_chosen"
                chosen_elements = container.find_all(id=chosen_class)
                select_info['has_chosen'] = len(chosen_elements) > 0
            except Exception as e:
                logger.warning(f"Error checking for chosen enhancement for {select_id}: {e}")
                select_info['has_chosen'] = False
        else:
            select_info['has_chosen'] = False
        
        result.append(select_info)
    
    return result

File: companion-app/forms/form_selects.js
/**
 * Extraction utilities for select dropdown controls.
 */
const { findLabelFor } = require('./form_common');
const { getLogger } = require('../utils/logger');

const logger = getLogger('form_extract_selects');

/**
 * Extract select dropdowns and their options
 * @param {Object} $ - Cheerio instance
 * @param {Object} container - Container element
 * @returns {Array} Array of select objects
 */
function extractSelects($, container) {
  const selects = [];
  
  $(container).find('select').each((i, select) => {
    const options = [];
    
    $(select).find('option').each((j, option) => {
      options.push({
        value: $(option).attr('value') || '',
        text: $(option).text().trim(),
        selected: $(option).attr('selected') !== undefined
      });
    });
    
    const selectInfo = {
      id: $(select).attr('id') || '',
      name: $(select).attr('name') || '',
      required: $(select).attr('required') !== undefined,
      disabled: $(select).attr('disabled') !== undefined,
      options: options
    };
    
    // Try to find label
    selectInfo.label = findLabelFor($, select);
    
    // Check for Chosen enhancement
    const selectId = $(select).attr('id');
    if (selectId) {
      try {
        // Use a safer approach to find the chosen container
        const chosenClass = `${selectId}_chosen`;
        const chosenElements = $(container).find(`#${chosenClass}`);
        selectInfo.has_chosen = chosenElements.length > 0;
      } catch (e) {
        logger.warning(`Error checking for chosen enhancement for ${selectId}: ${e.message}`);
        selectInfo.has_chosen = false;
      }
    } else {
      selectInfo.has_chosen = false;
    }
    
    selects.push(selectInfo);
  });
  
  return selects;
}

module.exports = { extractSelects };
